proposals/module-types/Explainer.md

```diff
@@ -0,0 +1,710 @@
+  (func (export "d") (result f32))
+)
+```
+Just as with module definitions, the above module type is actually an
+[abbreviation][import-abbrev] in the text format for:
+```wasm
+(module
+  (import "a" (memory 1 2))
+  (import "b" (func (param i32)))
+  (export "c" (table 1 funcref))
+  (export "d" (func (result f32)))
+)
+```
+which is the form we'll mostly use in this document.
+
+In WebAssembly there is also the separate concept of a module *instance*,
+which is the result of [instantiating][Module Instantiation] a module with
+imports. An instance is mostly just the module type with the imports removed
+(with the only future complication being [Type Imports] used in export
+signatures). For example, the above module, when instantiated, would have
+instance type:
+```wasm
+(instance
+  (export "b" (table 1 funcref))
+  (export "c" (func (result f32)))
+)
+```
+
+Just like function types, module and instance types can either be written
+"inline" or factored out into an explicit type definition that can be reused via
+`$identifier`. (Also, just like function types, the inline form is an
+[abbreviation][typeuse-abbrev] for an implicit type definition and use.) For
+example, an instance type can be defined:
+```wasm
+(type $WasiFile (instance
+  (export "read" (func (param i32 i32 i32) (result i32)))
+  (export "write" (func (param i32 i32 i32) (result i32)))
+))
+```
+
+In many examples shown below, type definitions are needed for *both* a module
+type and the instance type produced when that module type is instantiated. In
+such cases, to avoid duplicating all the exports, a new `(exports $instanceType)`
+type expression is added which can be used to define module types. For example,
+here is the type of a module which implements the above-defined `$WasiFile`
+interface via Win32 operations:
+```wasm
+(module
+  (import "Win32" "ReadFile" (func (param i32 i32 i32 i32) (result i32)))
+  (import "Win32" "WriteFile" (func (param i32 i32 i32 i32) (result i32)))
+  (exports $WasiFile)
+)
+```
+
+Lastly, just as the current [text format conventions recommend `.wat`][WAT] as
+the extension of a file that contains a module definition, this proposal
+includes a new recommendation for text files containing a bare module or
+instance type be suffixed `.wit`. `.wit` files can be used as part of the
+toolchain ecosystem for describing a module's interface without including its
+definition. For example, this can be used to generate compatible source-language
+declarations.
+
+Module and Instance types can also be used to describe the types of imports:
+
+
+### Instance Imports
+
+Just as a function, memory, table or global can be imported by specifying a
+function, memory, table or global type, modules or instances can be imported
+by specifying a module or instance type.
+
+A single instance import is practically equivalent to individually importing
+all of its exports. Indeed, instances have no distinct identity: instances are
```

Comment:
One thing I've been wondering while reading this and thinking about it, are instance imports necessary? I could see that in terms of a binary compression scheme of sorts it could be nice, but otherwise I'm not sure whether it's actually necessary to specify instance imports. For example you can always desugar an instance import into improting all the individual items.

I also think that instance imports lose some expressivity over individual imports in some cases. For example taking the virtualization example, if a module imports both `wasi_file` and `wasi_net` you might want to virtualize both of them. If the module imports two instances (one for `wasi_file` and one for `wasi_net`), you might have only one instance `my_virtual_file_and_net` which doesn't fit into that hole. This could also be solved with a "make an instance from this list of exports" instructions, but I wanted to bring this up here as a thought at least.

Reply:
I am also confused about when to import a module and when to import an instance. Is it the general rule that libraries import instances because they don't want to control instantiation and applications import modules because they do want to control instantiation?

Reply:
No, I don't think instance imports are strictly necessary.  I think they're nicer though:
* they group together the exports of an interface, making it easier to pass them around as a single unit between modules
* when listing the imports of a module, you get to see that a particular named interface is imported (once)

In the example you gave, `my_virtual_file_and_net` would be a subtype of both `wasi_file` and `wasi_net`, so you could pass it for both w/o trouble.  If necessary, local module+instance definitions can be used to do more fine-grained mixing and matching of fields.

@tlively Yes, (shared-everything) libraries import instances b/c they need to *share state* (and instance = state).  Programs import modules b/c they want to *create their own local state*, disjoint from other programs that may want to share the same libraries.